src/hotspot/share/runtime/os.cpp

```diff
@@ -1857,3 +1845,72 @@ void os::naked_sleep(jlong millis) {
   }
   naked_short_sleep(millis);
 }
+
+
+////// Implementation of PageSizes
+
+void os::PageSizes::add(size_t pagesize) {
+  assert(is_power_of_2(pagesize), "pagesize must be a power of 2: " SIZE_FORMAT_HEX, pagesize);
+  _v |= pagesize;
+}
+
+bool os::PageSizes::contains(size_t pagesize) const {
+  assert(is_power_of_2(pagesize), "pagesize must be a power of 2: " SIZE_FORMAT_HEX, pagesize);
+  return (_v & pagesize) > 0;
+}
+
+size_t os::PageSizes::next_smaller(size_t pagesize) const {
+  assert(is_power_of_2(pagesize), "pagesize must be a power of 2: " SIZE_FORMAT_HEX, pagesize);
+  size_t v2 = _v & (pagesize - 1);
+  if (v2 == 0) {
+    return 0;
+  }
+  return round_down_power_of_2(v2);
+}
+
+size_t os::PageSizes::next_larger(size_t pagesize) const {
+  assert(is_power_of_2(pagesize), "pagesize must be a power of 2: " SIZE_FORMAT_HEX, pagesize);
+  if (pagesize == max_power_of_2<size_t>()) { // Shift by 32/64 would be UB
+    return 0;
+  }
+  int l = exact_log2(pagesize) + 1;
+  size_t v2 = _v >> l;
+  if (v2 == 0) {
+    return 0;
+  }
+  return (size_t)1 << (l + count_trailing_zeros(v2));
+}
+
+size_t os::PageSizes::largest() const {
+  const size_t max = max_power_of_2<size_t>();
+  if (contains(max)) {
+    return max;
+  }
+  return next_smaller(max);
+}
+
+size_t os::PageSizes::smallest() const {
+  assert(min_page_size() > 0, "Sanity");
+  return next_larger(min_page_size() / 2);
```

Comment:
I'm not sure why you use min_page_size() / 2 instead of another small enough power_of_2, say the smallest one:
```
return next_larger(1);
```
and thereby get rid of the dependency on the odd min_page_size().

Reply:
True. Earlier implementation had a lop in next_larger. But now it does not matter. 1 is fine.

Reply:
(Note that min_page_size should be really a constant, and named differently. I have this mentally marked as another small cleanup)